// ng-image-slider/esm5/lib/slider-custom-image/slider-custom-image.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
/** @type {?} */
var youtubeRegExp = /^.*(youtu.be\/|v\/|u\/\w\/|embed\/|watch\?v=|\&v=|\?v=)([^#\&\?]*).*/;
/** @type {?} */
var validFileExtensions = ['jpeg', 'jpg', 'gif', 'png'];
/** @type {?} */
var validVideoExtensions = ['mp4'];
var SliderCustomImageComponent = /** @class */ (function () {
    function SliderCustomImageComponent() {
        this.YOUTUBE = 'youtube';
        this.IMAGE = 'image';
        this.VIDEO = 'video';
        this.fileUrl = '';
        this.fileExtension = '';
        this.type = this.IMAGE;
        this.isVideo = false;
    }
    Object.defineProperty(SliderCustomImageComponent.prototype, "imageUrl", {
        // @inputs
        set: /**
         * @param {?} url
         * @return {?}
         */
        function (url) {
            if (url && typeof (url) === 'string') {
                this.fileUrl = url;
                this.fileExtension = url.replace(/^.*\./, '');
                /** @type {?} */
                var match = url.match(youtubeRegExp);
                if (match && match[2].length === 11) {
                    this.type = this.YOUTUBE;
                    this.fileUrl = "https://img.youtube.com/vi/" + match[2] + "/0.jpg";
                }
                else if (this.fileExtension && validFileExtensions.indexOf(this.fileExtension.toLowerCase()) > -1) {
                    this.type = this.IMAGE;
                }
                else if (this.fileExtension && validVideoExtensions.indexOf(this.fileExtension.toLowerCase()) > -1) {
                    this.type = this.VIDEO;
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SliderCustomImageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SliderCustomImageComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SliderCustomImageComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    SliderCustomImageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'custom-img',
                    template: "<div *ngIf=\"fileUrl\">\n    <img [ngClass]=\"{'image': type === IMAGE, 'youtube': type === YOUTUBE}\" *ngIf=\"type === IMAGE || type === YOUTUBE\" [src]=\"fileUrl\">\n    <video [class]=\"VIDEO\" *ngIf=\"type === VIDEO\">\n        <source [src]=\"fileUrl\" type=\"video/mp4\"> Your browser does not support the video tag.\n    </video>\n    <span *ngIf=\"type === YOUTUBE || type === VIDEO || isVideo\" class=\"youtube-icon\"></span>\n</div>"
                }] }
    ];
    /** @nocollapse */
    SliderCustomImageComponent.ctorParameters = function () { return []; };
    SliderCustomImageComponent.propDecorators = {
        imageUrl: [{ type: Input }],
        isVideo: [{ type: Input }]
    };
    return SliderCustomImageComponent;
}());
export { SliderCustomImageComponent };
if (false) {
    /** @type {?} */
    SliderCustomImageComponent.prototype.YOUTUBE;
    /** @type {?} */
    SliderCustomImageComponent.prototype.IMAGE;
    /** @type {?} */
    SliderCustomImageComponent.prototype.VIDEO;
    /** @type {?} */
    SliderCustomImageComponent.prototype.fileUrl;
    /** @type {?} */
    SliderCustomImageComponent.prototype.fileExtension;
    /** @type {?} */
    SliderCustomImageComponent.prototype.type;
    /** @type {?} */
    SliderCustomImageComponent.prototype.isVideo;
}
