// ngx-carousels/esm5/rotate-carousel/rotate-carousel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Output, EventEmitter, ContentChildren, QueryList, ViewChild, ElementRef } from '@angular/core';
import { timer } from 'rxjs';
import { RotateCarouselItemDirective } from './rotate-carousel-item.directive';
/** @type {?} */
var DEFAULT_WIDTH = '300px';
/** @type {?} */
var DEFAULT_HEIGHT = '200px';
/** @type {?} */
var DEFAULT_DEPTH = '510px';
/** @type {?} */
var DEFAULT_PERIOD = 3500;
/** @type {?} */
var FULL_ANGLE = 360;
var RotateCarouselComponent = /** @class */ (function () {
    function RotateCarouselComponent() {
        /**
         * Emit event of rotation start
         */
        this.rotationStart = new EventEmitter();
        /**
         * Emit event of rotation end
         */
        this.rotationEnd = new EventEmitter();
        /**
         * Width of content in carousel item
         */
        this.width = DEFAULT_WIDTH;
        /**
         * Height of content in carousel item
         */
        this.height = DEFAULT_HEIGHT;
        /**
         * Depth of the center of carousel
         */
        this.depth = DEFAULT_DEPTH;
        /**
         * Period of carousel rotation
         */
        this.period = DEFAULT_PERIOD;
        /**
         * Direction of rotation
         */
        this.direction = 'forward';
        this._activeIndex = 0;
    }
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.calcItemsAngle();
        this.setupRotation();
        //this.calcHeight();
    };
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.setupRotation();
        //this.calcHeight();
    };
    //***************************************************************************************************************
    //-Setup---------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    //***************************************************************************************************************
    //-Setup---------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.calcItemsAngle = 
    //***************************************************************************************************************
    //-Setup---------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        this._angleStep = FULL_ANGLE / this.rotateItems.length;
        /** @type {?} */
        var angle = 0;
        this.rotateItems.forEach(function (x) {
            x.angle = angle;
            angle += _this._angleStep;
        });
    };
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.setupRotation = /**
     * @return {?}
     */
    function () {
        if (!Number.isNaN((/** @type {?} */ (this.period)) % 1))
            this._period = parseInt((/** @type {?} */ (this.period)));
        else {
            /** @type {?} */
            var periodStr = (/** @type {?} */ (this.period));
            // split string like '1003ms' to '1000' and 'ms'
            /** @type {?} */
            var values = periodStr.replace(/\'/g, '').split(/([0-9]*\.?[0-9])/);
            this._period = values[2] == 's' ? parseInt(values[1]) * 1000 : parseInt(values[1]);
        }
        if (this._rotationTimerSubscription)
            this._rotationTimerSubscription.unsubscribe();
        this._rotationTimerSubscription = this.setRotationTimer();
    };
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.calcHeight = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var maxH = Math.max.apply(Math, tslib_1.__spread([null], this.rotateItems.map(function (x) { return x.templateRef.elementRef.nativeElement.clientHeight; })));
        if (!this.height)
            this.height = maxH + 'px';
    };
    //***************************************************************************************************************
    //-Events--------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    //***************************************************************************************************************
    //-Events--------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.onRotationStart = 
    //***************************************************************************************************************
    //-Events--------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    function () {
        this.rotationStart.emit(this._activeIndex);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    RotateCarouselComponent.prototype.onRotateEnd = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this._activeIndex == 0) {
            this.carouselInner.nativeElement.style.transition = 'none';
            this.carouselInner.nativeElement.style.transform = 'rotateY(0deg)';
        }
        this.rotationEnd.emit(this._activeIndex);
    };
    //***************************************************************************************************************
    //-Style-Visibility-Availability---------------------------------------------------------------------------------
    //***************************************************************************************************************
    //***************************************************************************************************************
    //-Style-Visibility-Availability---------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.getCarouselStyle = 
    //***************************************************************************************************************
    //-Style-Visibility-Availability---------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var width = this.width;
        /** @type {?} */
        var height = this.height;
        /** @type {?} */
        var depth = this.depth;
        return { width: width, height: height, transform: 'translateZ(-' + depth + ')' };
    };
    /**
     * @param {?} item
     * @return {?}
     */
    RotateCarouselComponent.prototype.getItemStyle = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        /** @type {?} */
        var depth = this.depth;
        /** @type {?} */
        var style = { transform: 'rotateY(' + item.angle + 'deg) translateZ(' + depth + ')' };
        return style;
    };
    //***************************************************************************************************************
    //-Methods-------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    //***************************************************************************************************************
    //-Methods-------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @param {?} indexPath
     * @return {?}
     */
    RotateCarouselComponent.prototype.selectAt = 
    //***************************************************************************************************************
    //-Methods-------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @param {?} indexPath
     * @return {?}
     */
    function (indexPath) {
        this._rotationTimerSubscription.unsubscribe();
        this.carouselInner.nativeElement.style.transform = "rotateY(" + indexPath * this._angleStep + "deg)";
        this._rotationTimerSubscription = this.setRotationTimer();
    };
    //***************************************************************************************************************
    //-Help----------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    //***************************************************************************************************************
    //-Help----------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    RotateCarouselComponent.prototype.setRotationTimer = 
    //***************************************************************************************************************
    //-Help----------------------------------------------------------------------------------------------------------
    //***************************************************************************************************************
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        return timer(0, this._period)
            .subscribe(function (i) {
            _this._activeIndex = i % _this.rotateItems.length;
            /** @type {?} */
            var angle;
            if (_this._activeIndex == 1)
                _this.carouselInner.nativeElement.style.transition = "transform " + _this._period / 2 + "ms ease-in-out";
            if (i > 0 && _this._activeIndex == 0)
                angle = _this.rotateItems.length * _this._angleStep * (_this.direction == 'forward' ? -1 : 1);
            else
                angle = _this._activeIndex * _this._angleStep * (_this.direction == 'forward' ? -1 : 1);
            _this.carouselInner.nativeElement.style.transform = "rotateY(" + angle + "deg)";
            _this.onRotationStart();
        });
    };
    RotateCarouselComponent.decorators = [
        { type: Component, args: [{
                    selector: 'rotate-carousel',
                    template: "<figure class=\"carousel\">\r\n    <div class=\"carousel__content\" [ngStyle]=\"getCarouselStyle()\">\r\n        <div #carouselInner class=\"carousel__inner\" (transitionend)=\"onRotateEnd($event)\">\r\n            <ng-template ngFor let-item [ngForOf]=\"rotateItems\">\r\n                <div class=\"carousel__item\" [ngStyle]=\"getItemStyle(item)\">\r\n                    <ng-template [ngTemplateOutlet]=\"item.templateRef\"></ng-template>\r\n                </div>\r\n            </ng-template>\r\n        </div>\r\n    </div>\r\n</figure>",
                    styles: [":host{width:100%}.carousel{display:flex;justify-content:center;-webkit-perspective:1000px;perspective:1000px}.carousel__content{-webkit-transform:translateZ(-40vw);transform:translateZ(-40vw);-webkit-transform-style:preserve-3d;transform-style:preserve-3d}.carousel__inner{height:100%;-webkit-transform-style:preserve-3d;transform-style:preserve-3d}.carousel__item{display:flex;align-items:center;position:absolute;height:100%;width:100%}.carousel__item img{height:100%;width:100%;-o-object-fit:cover;object-fit:cover;-o-object-position:center;object-position:center;border-radius:15px}"]
                }] }
    ];
    RotateCarouselComponent.propDecorators = {
        rotateItems: [{ type: ContentChildren, args: [RotateCarouselItemDirective,] }],
        carouselInner: [{ type: ViewChild, args: ['carouselInner',] }],
        rotationStart: [{ type: Output }],
        rotationEnd: [{ type: Output }],
        width: [{ type: Input }],
        height: [{ type: Input }],
        depth: [{ type: Input }],
        period: [{ type: Input }],
        direction: [{ type: Input }]
    };
    return RotateCarouselComponent;
}());
export { RotateCarouselComponent };
if (false) {
    /** @type {?} */
    RotateCarouselComponent.prototype.rotateItems;
    /** @type {?} */
    RotateCarouselComponent.prototype.carouselInner;
    /**
     * Emit event of rotation start
     * @type {?}
     */
    RotateCarouselComponent.prototype.rotationStart;
    /**
     * Emit event of rotation end
     * @type {?}
     */
    RotateCarouselComponent.prototype.rotationEnd;
    /**
     * Width of content in carousel item
     * @type {?}
     */
    RotateCarouselComponent.prototype.width;
    /**
     * Height of content in carousel item
     * @type {?}
     */
    RotateCarouselComponent.prototype.height;
    /**
     * Depth of the center of carousel
     * @type {?}
     */
    RotateCarouselComponent.prototype.depth;
    /**
     * Period of carousel rotation
     * @type {?}
     */
    RotateCarouselComponent.prototype.period;
    /**
     * Direction of rotation
     * @type {?}
     */
    RotateCarouselComponent.prototype.direction;
    /** @type {?} */
    RotateCarouselComponent.prototype._period;
    /** @type {?} */
    RotateCarouselComponent.prototype._angleStep;
    /** @type {?} */
    RotateCarouselComponent.prototype._rotationTimerSubscription;
    /** @type {?} */
    RotateCarouselComponent.prototype._activeIndex;
}
